### PR TITLE
[MNG-7960] Artifact collection filtering

### PR DESCRIPTION
**Pull request by cstamas**
Make Maven more friendly (and tunable) when ranges are being used. There new options are merely affecting range processing, allowing different narrowing strategies. Relies on latest Resolver alpha-6 features.

---

https://issues.apache.org/jira/browse/MNG-7960

---

Demo: https://gist.github.com/cstamas/e7c72ac2499f41f632fc4dc8290c1157

### CODE REVIEW

**Comment by cstamas** [comment]
This PR depends on new Resolver alpha release.

File: maven-core/src/main/java/org/apache/maven/internal/aether/DefaultRepositorySystemSessionFactory.java [hunk 1]
```diff
@@ -406,6 +443,75 @@ public SessionBuilder newRepositorySessionBuilder(MavenExecutionRequest request)
...
+            List<String> expressions = Arrays.stream(filterExpression.split(","))
+                    .filter(s -> s != null && !s.trim().isEmpty())
+                    .collect(Collectors.toList());
+            for (String expression : expressions) {
+                if ("h".equals(expression)) {
+                    filters.add(new HighestVersionFilter());
+                } else if (expression.startsWith("h(") && expression.endsWith(")")) {
+                    int num = Integer.parseInt(expression.substring(2, expression.length() - 1));
+                    // MRESOLVER-450
+                    // filters.add(new HighestVersionFilter(num));
```
Comment (gnodet) [inline maven-core/src/main/java/org/apache/maven/internal/aether/DefaultRepositorySystemSessionFactory.java:459]: I think it's https://github.com/apache/maven-resolver/pull/401 / MRESOLVER-460.
This PR should be moved to draft to avoid anyone forgetting it's not finished and we need next resolver alpha.

**Review by gnodet — CHANGES_REQUESTED** [review]
This PR is not in a mergeable state yes and requires next resolver alpha.

**Review by gnodet — APPROVED** [review]

File: pom.xml [hunk 2]
```diff
@@ -176,7 +176,7 @@ under the License.
     <plexusInterpolationVersion>1.26</plexusInterpolationVersion>
     <plexusTestingVersion>1.0.0</plexusTestingVersion>
     <plexusXmlVersion>4.0.1</plexusXmlVersion>
-    <resolverVersion>2.0.0-alpha-5</resolverVersion>
+    <resolverVersion>2.0.0-alpha-6</resolverVersion>
```
Comment (gnodet) [inline pom.xml:179]: Should that be in a separate commit for clarity and release notes ?

Comment (cstamas) [reply pom.xml:179]: There is a separate PR just for the update https://github.com/apache/maven/pull/1373 (and removal of some deprecation warnings in tests)
